Caltech-IPAC/firefly
# DM-11814: Use embedded DB in table data support

https://jira.lsstcorp.org/browse/DM-11814

This pull request replaces our ipac-table based table support with database backend.
By default, HyperSQL(HSQLDB) is used.  However, you can change the default by setting DbAdapter.type in common.prop.
We currently support h2(H2), hsql(HSQLDB), and sqlite(SQLite).

This conversion is almost complete.  There are a few known bugs I've not fixed.  It will be before this PR merged.
- catalog overlay lost on coverage after filtering.
- histogram chart not showing.
- heatmaps in demo/ffapi-highlevel-charttest.html does not show all three on first load, but will eventually.

To test:  build and deploy firefly on this branch.
Every table displayed on the app is now backed by a database in the backend.
You'll find the database files in the temp_files directory.  The file extension matches the database type.

To test the different database types, 
go to http://localhost:8080/firefly/firefly-dev.html
 - click "Data Sets: Catalogs & Images"
 - enter a file path under "Source Table"
 - select the database engine to use under "dbType"
 - click "Search",  you will see the selected dbType appended to your table's title.
 
To see how column expression can be used:
 - goto http://localhost:8080/firefly/demo/ffapi-highlevel-charttest.html
 - at the very end, there's a second table, titled "table with expressions". 
 - this table is created by selecting `ra + dec as "radec"', ra, dec, ln(dec) as lnofdec, power(ra, 2) as ra_sq`
 - note the double-quotes around radec.  that keeps it case sensitive.  otherwise, most database will stored it as uppercase.
 
Useful tools:
- To see only the performance printout in the log files.  
  - This will show the time it took to complete each individual steps.
  - It will also show when a search processor is not going through the optimized path.  You see it identified with "<< slow " follow by its name.
  - >  tail -f ${CATALINA_HOME}/logs/*[a-zA-Z].log ${CATALINA_HOME}/logs/*.out | grep -E --line-buffered "<<|StopWatch" 

- To access the database files using a jdbc client.
  - goto http://localhost:8080/firefly/admin/db/
  - user/password is same as the tomcat manager's
  - select the database type in the first drop-down box.
  - insert the database file you want to work with in "JDBC URL"
  - click connect.
  - The UI is similar to other jdbc client and is pretty much self-explanatory 
 

Files to focus on:
- edu.caltech.ipac.firefly.server.query.DbProcessor
  - New SearchProcessor base on a database backend.
- edu.caltech.ipac.firefly.server.query.DbProcessorWrapper
  - A wrapper SearchProcessor to adapt the ipac-table based processors to the new DbProcessor framework.  This uses the original processor to create the data, but store it in a database.
- edu.caltech.ipac.firefly.server.db.DbAdapter
  - An interface to abstract the implementations of various database engines.
- src/firefly/html/demo/ffapi-highlevel-charttest.html
  - A sample of how column expressions can be used in a table.
